adoc/chapters/programming_interface.adoc

```diff
@@ -17945,43 +17945,77 @@ device_event(___unspecified___)
-[code]#unsigned long#,
-[code]#long long#,
-[code]#unsigned long long#, [code]#float# or
-[code]#double#.  For floating-point types, the member functions of the
-[code]#atomic_ref# class may be emulated, and may use a different
-floating-point environment to those defined by
-[code]#info::device::single_fp_config# and
-[code]#info::device::double_fp_config# (i.e. floating-point atomics may
-use different rounding modes and may have different exception behavior).
+The [code]#sycl::atomic_ref# class has a template parameter
+[code]#DefaultScope#, which allows the application to define a default memory
+scope for the atomic operations.  Each member function also provides an
+optional parameter that allows the application to override this default.
+
+The [code]#sycl::atomic_ref# class also has a template parameter
+[code]#AddressSpace#, which allows the application to make an assertion about
+the address space of the object of type [code]#T# that it references.  The
+default value for this parameter is
+[code]#access::address_space::generic_space#, which indicates that the object
+could be in either the global or local address spaces.  If the application
+knows the address space, it can set this template parameter to either
+[code]#access::address_space::global_space# or
+[code]#access::address_space::local_space# as an assertion to the
+implementation.  Specifying the address space via this template parameter may
+allow the implementation to perform certain optimizations.  Specifying an
+address space that does not match the object's actual address space results in
+undefined behavior.
+
+The template parameter [code]#T# must be one of the following types:
+
+* [code]#int#,
+* [code]#unsigned int#,
+* [code]#long#,
+* [code]#unsigned long#,
+* [code]#long long#,
+* [code]#unsigned long long#,
+* [code]#float#, or
+* [code]#double#.
+
+In addition, the type [code]#T# must satisfy one of the following conditions:
+
+* [code]#sizeof(T) == 4#, or
+* [code]#sizeof(T) == 8# and the code containing this [code]#atomic_ref# was
+  submitted to a device that has [code]#aspect::atomic64#.
+
+For floating-point types, the member functions of the [code]#atomic_ref# class
+may be emulated, and they may use a different floating-point environment from
+those defined by [code]#info::device::single_fp_config# and
+[code]#info::device::double_fp_config# (i.e. floating-point atomics may use
+different rounding modes and may have different exception behavior).
+
+If the code using [code]#sycl::atomic_ref# is submitted to a device that has
+[code]#aspect::usm_atomic_host_allocations# or
+[code]#aspect::usm_atomic_shared_allocations#, then atomic operations in device
+code (using [code]#sycl::atomic_ref#) are guaranteed to be atomic with respect
+to atomic operations performed in host code (using [code]#std::atomic_ref#),
+when they operate on the same object and when the memory scope is
+[code]#memory_scope::system#.
```

Comment:
I know this PR is focused on `atomic_ref`, but should we add similar clarifications for the interaction between `sycl::atomic_fence` and `std::atomic_thread_fence`, and the interaction between fences and atomic references?

One simple way to do this might be to lift this paragraph to the "Synchronization and atomics" section and add something like:

> Additionally, atomic operations in device code are guaranteed to be able to synchronize with fence operations performed in host code (using [code]#std::atomic_thread_fence#), and atomic operations in host code are guaranteed to be able to synchronize with fence operations performed in device code (using [code]#sycl::atomic_fence#), when the memory scope is [code]#memory_scope::system#.

I suggest "guaranteed to be able to" rather than "guaranteed to" in the above because fences only synchronize with atomics and other fences under certain conditions (see [atomics.fences](http://eel.is/c++draft/atomics.fences)).

Reply:
Yes, something like this probably makes sense.  It seems weird, though, to put that description at the start of the [section 4.15](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_synchronization_and_atomics) before we talk about either `sycl::atomic_fence` or `sycl::atomic_ref`.  Maybe it would be better to add a new section 4.15.5 "Synchronization between host and device code" where we describe this.

Reply:
I agree, a new section would be better.

Reply:
My proposed wording is in 5ef7fe95aee51656a85cb7025254892503f3604a.

Reply:
I think this looks very good.  I'm worried I'm being a little pedantic, but I think we should consider dropping the "share the same atomic object _M_" bit from the last three paragraphs.

I don't think it's obvious what it means to "share the same atomic object _M_", because a fence doesn't directly operate on any memory objects.

I suggest a simpler "if the fence operations follow the rules for fence synchronization defined in the {cpp} core language.".  This still requires certain orderings of fence operations and atomic operations, but which orderings lead to synchronization is covered entirely by C++.

Reply:
I'd like to keep that because it helps clarify why these fence operations only provide a guarantee of synchronization between the host and device when the device supports `aspect::usm_atomic_host_allocations` or `aspect::usm_atomic_shared_allocations`.  (Since C++ defines the fence operations in terms of atomic operations on some shared object _M_, there is only a synchronization guarantee when it is possible to have some shared atomic object _M_.)

I agree that the wording would be unclear to a casual reader because you would have to read the C++ standard to know about this relationship between fence operations and a shared atomic object.  I tried to fix this in 7d52276711524cae6596e3623cf8885426175c39 by clarifying that this relationship exists.